### PR TITLE
Fix bug: LandMine is not exploding to enemies.

### PR DESCRIPTION
**Pull request by Toneyisnow**
Fix bug: LandMine is not exploding to enemies.

The reason is,
the mine has attribute hidden=true;
when enemy unit moves, the code in BattleInfo.cpp MoveUnit() (line 817) will update the revealed to true;
then in the CGameHandler.cpp handleDamageFromObstacle() (line 4846) is checking , and the condition battleIsObstacleVisibleForSide() will return true, so the effect will not be triggerred.

Resolution:
1. Remove the "revealed=true" in moveUnit(), and in handleDamageFromObstacle, remove the "const" restrict for obstacle, and then update revealed to true;
2. After the takeDamage function, add a pack "BattleObstaclesChanged" to update the obstacle to be "revealed=true".

### CODE REVIEW

File: server/CGameHandler.cpp [hunk 1]
```diff
@@ -4858,6 +4859,25 @@ bool CGameHandler::handleDamageFromObstacle(const CStack * curStack, bool stackI
 
 					if(oneTimeObstacle)
 						removeObstacle(*obstacle);
```
Comment (alexvins) [inline server/CGameHandler.cpp]: Put `{}` around because else clause in not one line.

File: lib/battle/IBattleState.h [hunk 2]
```diff
@@ -87,5 +87,6 @@ class DLL_LINKAGE IBattleState : public IBattleInfo
 	virtual void setWallState(int partOfWall, si8 state) = 0;
 
 	virtual void addObstacle(const ObstacleChanges & changes) = 0;
+	virtual void updateObstacle(const ObstacleChanges& changes) = 0;
```
Comment (alexvins) [inline lib/battle/IBattleState.h]: - This change requires `BattleStateMock` to be updated.
- Put spaces around `&`

Comment (Toneyisnow) [reply lib/battle/IBattleState.h]: Updated.

File: server/CGameHandler.cpp (hunk 1 above)
Comment (Toneyisnow) [reply server/CGameHandler.cpp]: Updated.

File: server/CGameHandler.cpp [hunk 3]
```diff
@@ -51,6 +51,7 @@
 #ifndef _MSC_VER
 #include <boost/thread/xtime.hpp>
 #endif
+#include <serializer\JsonSerializer.h>
```
Comment (alexvins) [inline server/CGameHandler.cpp]:
```suggestion
#include "../lib/serializer/JsonSerializer.h"
```

Comment (Toneyisnow) [reply server/CGameHandler.cpp]: Oh, forgot this one

Comment (Toneyisnow) [reply server/CGameHandler.cpp]: Oh, forgot this one

**Pull request merged**